lib/streamlit/elements/widgets/number_input.py

```diff
@@ -73,47 +75,82 @@ def deserialize(
+    # If "max_value: int, value=None" is given and all other numerical inputs are "int"s or not provided, return "int | None"
     @overload
     def number_input(
         self,
         label: str,
-        min_value: Number | None = None,
-        max_value: Number | None = None,
-        value: None = None,
-        step: Number | None = None,
-        format: str | None = None,
-        key: Key | None = None,
-        help: str | None = None,
-        on_change: WidgetCallback | None = None,
-        args: WidgetArgs | None = None,
-        kwargs: WidgetKwargs | None = None,
-        *,  # keyword-only arguments:
-        placeholder: str | None = None,
-        disabled: bool = False,
-        label_visibility: LabelVisibility = "visible",
-    ) -> Number | None:
-        pass
+        min_value: int | None = None,
+        *,
+        max_value: int,
+        value: _IntOrNone | Literal["min"] = "min",
+        step: int | None = None,
+        format: str | None = None, key: Key | None = None, help: str | None = None, on_change: WidgetCallback | None = None, args: WidgetArgs | None = None, kwargs: WidgetKwargs | None = None, placeholder: str | None = None, disabled: bool = False, label_visibility: LabelVisibility = "visible"
+    ) -> int | _IntOrNone:
+        ...
+
+    # If "value=int" is given and all other numerical inputs are "int"s or not provided, return "int"
+    @overload
+    def number_input(
+        self,
+        label: str,
+        min_value: int | None = None,
+        max_value: int | None = None,
+        *,
+        value: int,
+        step: int | None = None,
+        format: str | None = None, key: Key | None = None, help: str | None = None, on_change: WidgetCallback | None = None, args: WidgetArgs | None = None, kwargs: WidgetKwargs | None = None, placeholder: str | None = None, disabled: bool = False, label_visibility: LabelVisibility = "visible"
+    ) -> int:
+        ...
+
+    # If "step=int" is given and all other numerical inputs are "int"s or not provided (value optionally being "min"), return "int"
+    # If "step=int, value=None" is given and all other numerical inputs are "int"s or not provided, return "int | None"
+    @overload
+    def number_input(
+        self,
+        label: str,
+        min_value: int | None = None,
+        max_value: int | None = None,
+        value: _IntOrNone | Literal["min"] = "min",
+        *,
+        step: int,
+        format: str | None = None, key: Key | None = None, help: str | None = None, on_change: WidgetCallback | None = None, args: WidgetArgs | None = None, kwargs: WidgetKwargs | None = None, placeholder: str | None = None, disabled: bool = False, label_visibility: LabelVisibility = "visible"
+    ) -> int | _IntOrNone:
+        ...
+
+    # If all numerical inputs are floats (with value optionally being "min") or are not provided, return "float"
```

Comment:
nit: couldn't we also add an overload for the situation that everything is float -> return float?